New page: overflow values (#26395)
mdn/content: files/en-us/web/css/overflow_value/index.md

```css hidden
pre {
  border: 2px dashed red;
  margin-bottom: 3em;
}

::before {
  font-weight: bold;
  color: white;
  background: crimson;
  display: inline-block;
  width: 100%;
  padding: 3px 5px;
  box-sizing: border-box;
}
```

```css
pre {
  block-size: 100px;
  inline-size: 295px;
}

pre:nth-of-type(1) {
  overflow: hidden;
}
pre:nth-of-type(1)::before {
  content: "hidden: ";
}

pre:nth-of-type(2) {
  overflow: clip;
  overflow-clip-margin: 1em;
}
pre:nth-of-type(2)::before {
  content: "clip: ";
}

pre:nth-of-type(3) {
  overflow: scroll;
}
pre:nth-of-type(3)::before {
  content: "scroll: ";
}

pre:nth-of-type(4) {
  overflow: auto;
}
pre:nth-of-type(4)::before {
  content: "auto: ";
}

pre:nth-of-type(5) {
  overflow: clip;
  overflow: overlay;
  overflow-clip-margin: 3em;
}
pre:nth-of-type(5)::before {
  content: "overlay (or clip if not supported): ";
}

pre:nth-of-type(6) {
  overflow: visible;
}
pre:nth-of-type(6)::before {
  content: "visible: ";
}
```

### Result

> {{EmbedLiveSample("Examples", "500", "600")}}

Author: ```suggestion
{{EmbedLiveSample("Examples", "500", "800")}}
```

---
title: <overflow>
slug: Web/CSS/overflow_value

> page-type: css-type

Author: for now, we are adding this as a regular value type. We can change if we think it doesn't work live. 
For the slug, other CSS data types use "value" not "values.

> browser-compat: css.types.overflow

Author: we're pretending this is a regular data type for now: it's just enumerated rather than functional or numeric like the other ones

---

{{CSSRef}}

> The **`<overflow>`** {{glossary("enumerated")}} values is a set of keyword values used in the {{cssxref("overflow-x")}}, {{cssxref("overflow-y")}}, {{cssxref("overflow-inline")}} and {{cssxref("overflow-block")}} properties, as well as the {{cssxref("overflow")}} shorthand property. These properties apply to block containers, flex containers, and grid containers.

Author: ```suggestion
The **`<overflow>`** {{glossary("enumerated")}} value is a set of keyword values used in the {{cssxref("overflow-x")}}, {{cssxref("overflow-y")}}, {{cssxref("overflow-inline")}} and {{cssxref("overflow-block")}} properties, as well as the {{cssxref("overflow")}} shorthand property. These properties apply to block containers, flex containers, and grid containers.
```

> The **`<overflow>`** {{glossary("enumerated")}} value type defines the set of overflow keyword values used in the {{cssxref("overflow-x")}}, {{cssxref("overflow-y")}}, {{cssxref("overflow-inline")}} and {{cssxref("overflow-block")}} properties, as well as the {{cssxref("overflow")}} shorthand property. These properties apply to block containers, flex containers, and grid containers.

Reviewer: ```suggestion
The **`<overflow>`** {{glossary("enumerated")}} value type represents the keyword values for the {{cssxref("overflow-block")}}, {{cssxref("overflow-inline")}}, {{cssxref("overflow-x")}}, and {{cssxref("overflow-y")}} longhand properties and the {{cssxref("overflow")}} shorthand property. These properties apply to block containers, flex containers, and grid containers.
```

Does the [Enumerated](https://developer.mozilla.org/en-US/docs/Glossary/Enumerated) glossary page need to be updated to include a mention of enumerated values in CSS?

Author: Good call. I think we're ok with the current definition: "In computer science, an enumerated type is a data type consisting of a limited set of named values.". It'ss enough of a definition for CSS. Enumerated in HTML and JS has special meaning in addition to that line.

## Syntax

```css
<overflow> = visible | hidden | clip | scroll | auto
```

## Values

> The `<overflow>` data type is specified using one of the values listed below.

Reviewer: ```suggestion
The `<overflow>` enumerated value type is specified using one of the values listed below.
```

> - `visible`
>   - : Overflow content is not clipped and may be visible outside the element's padding box. The element box is not a {{glossary("scroll container")}}. This is the default value of all the properties which have the `<overflow>` as the data type.
> - `hidden`
>   - : Overflow content is clipped at the element's padding box. There are no scroll bars, and the clipped content is not visible (i.e., clipped content is hidden), but the content still exists. User agents do not add scroll bars and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ be scrolled programmatically (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method). The content can also be scrolled via keyboard interaction; arrows enable scrolling thru the content and tabbing to a focusable element within the hidden content scrolls that focused element into view. The element box on which this value is set is a {{glossary("scroll container")}}.
> - `clip`
>   - : Overflow content is clipped at the element's _overflow clip edge_ that is defined using the [`overflow-clip-margin`](/en-US/docs/Web/CSS/overflow-clip-margin) property. As a result, content overflows the element's padding box by the {{cssxref("&lt;length&gt;")}} value of `overflow-clip-margin` or by `0px` if not set. Overflow content outside the clipped region is not visible, user agents do not add a scroll bar, and programmatic scrolling is also not supported. No new [formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) is created.
> - `scroll`
>   - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view using scroll bars. User agents display scroll bars in both horizontal and vertical directions if only one value is set, whether or not any content is overflowing or clipped. The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. Printers may still print overflowing content. The element box on which this value is set is a {{glossary("scroll container")}}.
> - `auto`
>   - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view. Unlike `scroll`, user agents display scroll bars _only if_ the content is overflowing and hide scroll bars by default. If content fits inside the element's padding box, it looks the same as with `visible` but still establishes a new formatting context. The element box on which this value is set is a {{glossary("scroll container")}}.

Reviewer: ```suggestion
- `visible`
  - : Overflow content is not clipped and may be visible outside the element's padding box. The element box is not a {{glossary("scroll container")}}. This is the default value for all the properties that have the `<overflow>` enumerated value type.
- `hidden`
  - : Overflow content is clipped at the element's padding box. There are no scroll bars, and the clipped content is not visible (i.e., clipped content is hidden), but the content still exists. User agents do not add scrollbars and also do not allow users to view the content outside the clipped region by actions such as dragging on a touch screen or using the scroll wheel on a mouse. The content _can_ be scrolled programmatically (for example, by setting the value of the {{domxref("Element.scrollLeft", "scrollLeft")}} property or the {{domxref("Element.scrollTo", "scrollTo()")}} method). The content can also be scrolled via keyboard interaction; arrows enable scrolling through the content and tabbing to a focusable element within the hidden content enables scrolling the focused element into view. The element box on which this value is set is a scroll container.
- `clip`
  - : Overflow content is clipped at the element's _overflow clip edge_ that is defined using the [`overflow-clip-margin`](/en-US/docs/Web/CSS/overflow-clip-margin) property. As a result, content overflows the element's padding box by the {{cssxref("&lt;length&gt;")}} value of `overflow-clip-margin` or by `0px` if not set. Overflow content outside the clipped region is not visible, user agents do not add a scrollbar, and programmatic scrolling is also not supported. No new [formatting context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) is created.
- `scroll`
  - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view using scrollbars. User agents display scrollbars in both horizontal and vertical directions if only one value is set, whether or not any content is overflowing or clipped. The use of this keyword value, therefore, can prevent scrollbars from appearing and disappearing as content changes. Printers may still print overflowing content. The element box on which this value is set is a scroll container.
- `auto`
  - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view. Unlike `scroll`, user agents display scrollbars _only if_ the content is overflowing and hide scrollbars by default. If content fits inside the element's padding box, it looks the same as with `visible` but still establishes a new formatting context. The element box on which this value is set is a scroll container.
```
- I had added a space in "scroll bars" but after looking at some articles, I am now leaning towards the one word "scrollbar".
- In `scroll` description, we say "User agents display scrollbars in both horizontal and vertical directions if only one value is set" but we have not mentioned on this page that properties can accept two values, one for each direction.

> **Note:** The keyword value `overlay` is a legacy value alias for `auto`. With `overlay`, the scroll bars are drawn on top of the content instead of taking up space.

## Examples

> This example shows all the `<overflow>` values as values of the CSS {{cssxref("overflow")}} property.

Reviewer: ```suggestion
This example demos all the `<overflow>` enumerated values for the {{cssxref("overflow")}} property.
```

### HTML

> We include lyrics, using the HTML {{HTMLELement("pre")}} element for preformatted content. The same HTML is repeated multiple times. A link is included to enable testing the effects of keyboard focus on overflow and scroll behaviors.

Reviewer: ```suggestion
The HTML in this example contains some lyrics within the {{HTMLELement("pre")}} element. The HTML also contains a link text to enable testing the effects of keyboard focus on overflow and scroll behaviors. The same HTML code is repeated multiple times to show the effect of each `<overflow>` enumerated value. 
```
tabbing on link will not work on Firefox

I find a little fella who's cute and yellow and chubby
Rub-a-dub-dubby

<a href="#">Rubber Duckie</a>, you're so fine
And I'm lucky that you're mine
Rubber Duckie, I'm awfully fond of you
</pre>

<pre>&nbsp;
Oh, Rubber Duckie, you're the one
You make bath time lots of fun
Rubber Duckie, I'm awfully fond of you

Rubber Duckie, joy of joys
When I squeeze you, you make noise
Rubber Duckie, you're my very best friend, it's true

Oh, every day when I make my way to the tubby
I find a little fella who's cute and yellow and chubby
Rub-a-dub-dubby

<a href="#">Rubber Duckie</a>, you're so fine
And I'm lucky that you're mine
Rubber Duckie, I'm awfully fond of you
</pre>

<pre>&nbsp;
Oh, Rubber Duckie, you're the one
You make bath time lots of fun
Rubber Duckie, I'm awfully fond of you

Rubber Duckie, joy of joys
When I squeeze you, you make noise
Rubber Duckie, you're my very best friend, it's true

Oh, every day when I make my way to the tubby
I find a little fella who's cute and yellow and chubby
Rub-a-dub-dubby

<a href="#">Rubber Duckie</a>, you're so fine
And I'm lucky that you're mine
Rubber Duckie, I'm awfully fond of you
</pre>
```

### CSS

> We define the size of `<pre>` to ensure the content overflows its container in both the inline and block directions. We set a different `<overflow>` value in each example, and include generated content to help you identify which overflow value is being demonstrated:

Reviewer: ```suggestion
For the purpose of demonstration, the size of the `<pre>` element box has been defined to ensure that the content overflows its container in both the inline and block directions. A different `<overflow>` value is set for each of the repeating `<pre>` elements. For the `clip` value demonstration, a {{CSSXref("overflow-clip-margin")}} has been added.
```

```css hidden
pre {
  border: 2px dashed red;
  margin-bottom: 3em;
}

::before {
  font-weight: bold;
  color: white;
  background: crimson;
  display: inline-block;
  width: 100%;
  padding: 3px 5px;
  box-sizing: border-box;
}
```

```css
pre {
  block-size: 100px;
  inline-size: 295px;
}

pre:nth-of-type(1) {
  overflow: hidden;
}
pre:nth-of-type(1)::before {
  content: "hidden: ";
}

pre:nth-of-type(2) {
  overflow: clip;
  overflow-clip-margin: 1em;
}
pre:nth-of-type(2)::before {
  content: "clip: ";
}

pre:nth-of-type(3) {
  overflow: scroll;
}
pre:nth-of-type(3)::before {
  content: "scroll: ";
}

pre:nth-of-type(4) {
  overflow: auto;
}
pre:nth-of-type(4)::before {
  content: "auto: ";
}

pre:nth-of-type(5) {
  overflow: clip;
  overflow: overlay;
  overflow-clip-margin: 3em;
}
pre:nth-of-type(5)::before {
  content: "overlay (or clip if not supported): ";
}

pre:nth-of-type(6) {
  overflow: visible;
}
pre:nth-of-type(6)::before {
  content: "visible: ";
}
```

> We included {{CSSXref("overflow-clip-margin")}}, which only has an effect when used in conjunction with the `clip` value to demonstrate the clip effect.
> 

Reviewer: ```suggestion

```
Moved this up along with CSS code description

> ### Results

Reviewer: ```suggestion
### Result
```

> {{EmbedLiveSample("Examples", "500", "600")}}

Reviewer: You might want to increase the length of the iframe here to contain all the examples

> A link is included in each `<pre>` to demonstrate the effects of keyboard focus on overflow and scroll behaviors. Try tabbing thru all the links in the example. Notice that the `clip` example does not create a {{glossary("scroll container")}}, and the link does not come into view when the link is focused. The `visible` value, which has the link always in view, is also not a scroll container.

Reviewer: ```suggestion
To see the effect of keyboard focus on overflow and scroll behaviors, try tabbing through all the links in the example. Notice that the `clip` box does not create a scroll container, and the link does not come into view when the link is focused. The `visible` value, which has the link always in view, is also not a scroll container.
```

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - Properties that use this data type: {{cssxref("overflow-x")}}, {{cssxref("overflow-y")}}, {{cssxref("overflow-inline")}}, {{cssxref("overflow-block")}} and {{cssxref("overflow")}}

Reviewer: ```suggestion

```
The list has already been linked at the top of the page

- `scroll`
  - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view using scroll bars. User agents display scroll bars in both horizontal and vertical directions if only one value is set, whether or not any content is overflowing or clipped. The use of this keyword, therefore, can prevent scroll bars from appearing and disappearing as content changes. Printers may still print overflowing content. The element box on which this value is set is a {{glossary("scroll container")}}.
- `auto`
  - : Overflow content is clipped at the element's padding box, and overflow content can be scrolled into view. Unlike `scroll`, user agents display scroll bars _only if_ the content is overflowing and hide scroll bars by default. If content fits inside the element's padding box, it looks the same as with `visible` but still establishes a new formatting context. The element box on which this value is set is a {{glossary("scroll container")}}.

> > **Note:** The keyword value `overlay` is a legacy value alias for `auto`. With `overlay`, the scroll bars are drawn on top of the content instead of taking up space.

Reviewer 2: The `overlay` value *should* strictly behave the same as `auto` now per https://github.com/w3c/csswg-drafts/issues/8063#issuecomment-1431620007. However, according to [Firefox Bug 1817189](https://bugzilla.mozilla.org/show_bug.cgi?id=1817189), [WebKit Changeset 236341](https://trac.webkit.org/changeset/236341/webkit), and [Chrome Status](https://chromestatus.com/feature/5194091479957504), it seems that Chrome alone retains the legacy behavior.

Reviewer: In testing on Firefox 112, Chrome 112, and Safari 16, this behavior of `overlay`, where the scroll bars are drawn on top of the content, is found to be true. So we're calling out that difference from `auto`.

</pre>
```

### CSS

We define the size of `<pre>` to ensure the content overflows its container in both the inline and block directions. We set a different `<overflow>` value in each example, and include generated content to help you identify which overflow value is being demonstrated:

```css hidden
pre {
  border: 2px dashed red;
  margin-bottom: 3em;
}

::before {
  font-weight: bold;
  color: white;
  background: crimson;
  display: inline-block;
  width: 100%;
  padding: 3px 5px;
  box-sizing: border-box;
}
```

```css
pre {
  block-size: 100px;
  inline-size: 295px;
}

pre:nth-of-type(1) {
  overflow: hidden;
}
pre:nth-of-type(1)::before {
  content: "hidden: ";
}

pre:nth-of-type(2) {
  overflow: clip;
  overflow-clip-margin: 1em;
}
pre:nth-of-type(2)::before {
  content: "clip: ";
}

pre:nth-of-type(3) {
  overflow: scroll;
}
pre:nth-of-type(3)::before {
  content: "scroll: ";
}

pre:nth-of-type(4) {
  overflow: auto;
}
pre:nth-of-type(4)::before {
  content: "auto: ";
}

pre:nth-of-type(5) {
  overflow: clip;
  overflow: overlay;
  overflow-clip-margin: 3em;
}
pre:nth-of-type(5)::before {
  content: "overlay (or clip if not supported): ";

> }

Reviewer 2: …and this block, along with one of the `<p>` elements above, could be removed.
